Make deps directory persistent over upgrades

MartinHjelmare opened the pull request:
## Description:
* Install requirements to venv with pip if venv is active.
* Install requirements with --user option if venv is not active.
* Set PYTHONUSERBASE environment variable to deps directory, when installing with --user option.
* Reset --prefix option to workaround incompatibility when installing with --user option. This requires pip version 8.0.0 or greater.
* Require pip version 8.0.3.
* Do not delete deps directory on home assistant upgrade.
* Fix local lib mount and check package exist.
* Add and update tests.
* Fix upgrade from before version 0.49, by removing deps directory one final time.

**Related issue (if applicable):**
fixes #3807 
Background: #5892

## Checklist:
If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

MartinHjelmare committed:
Use pip install --user if venv not active

* Set PYTHONUSERBASE to deps directory, when installing with --user
  option.
* Reset --prefix option to workaround incompatability when installing
  with --user option. This requires pip version 8.0.0 or greater.
* Require pip version 8.0.3.
* Do not delete deps directory on home assistant upgrade.
* Fix local lib mount and check package exist.

MartinHjelmare committed:
Update and add tests

MartinHjelmare committed:
Fix upgrade from before version 0.46

MartinHjelmare commented:
It would be good to have this tested in a Windows environment before merging.

pvizeli commented on homeassistant/bootstrap.py:
```diff
@@ -281,6 +282,12 @@ def mount_local_lib_path(config_dir: str) -> str:
     deps_dir = os.path.join(config_dir, 'deps')
-    if deps_dir not in sys.path:
-        sys.path.insert(0, os.path.join(config_dir, 'deps'))
+    env = os.environ.copy()
```
Why you copy this?

MartinHjelmare replied:
I didn't want to modify the environment permanently. I'm not sure if that matters though, but I thought it safer.

balloob replied:
It's indeed safer to not modify it.

balloob commented on homeassistant/bootstrap.py:
```diff
@@ -281,6 +282,12 @@ def mount_local_lib_path(config_dir: str) -> str:
+    env['PYTHONUSERBASE'] = os.path.abspath(deps_dir)
+    args = [sys.executable, '-m', 'site', '--user-site']
+    process = Popen(args, stdin=PIPE, stdout=PIPE, stderr=PIPE, env=env)
+    stdout, _ = process.communicate()
```
This is no longer async friendly.

balloob replied:
We should probably move this to the `util.package`: `get_user_site`. I actually wonder if we can't fetch this while running?

MartinHjelmare replied:
Can we use this instead: https://docs.python.org/3/library/asyncio-subprocess.html#create-a-subprocess-low-level-api-using-subprocess-popen

pvizeli replied:
look into `image_processing.openalpr_local`. I use async subprocess there and write also a mock for tests.

balloob commented on homeassistant/setup.py:
```diff
@@ -76,6 +77,10 @@ def _async_process_requirements(hass: core.HomeAssistant, name: str,
 
     def pip_install(mod):
         """Install packages."""
+        if hasattr(sys, 'real_prefix'):
```
Let's extract if we are in a virtual environment into a helper function.

balloob commented:
This is awesome btw!

balloob commented:
Hey Martin, any update on this ?

MartinHjelmare commented:
I've switched to async subprocess for getting the user site and broke out two new functions for that and moved them two package util. But I had troubles with the async tests for this. I think I can fix that now after seeing how you tested the python_script addition. I'm not super confident about how to structure the functions. I was thinking we need two, one sync and one async. The sync one is needed for the scripts package.

I'll try to fix the tests and then I can push for more input from reviewers. I've been busy the last couple of weeks and currently dodging a cold. :face_with_thermometer:

balloob commented:
No problem!

You can always decide just to expose an async version. In the script package we can wrap it in a `run_coroutine_threadsafe`.

MartinHjelmare committed:
Extract function to get user site

* Add function(s) to package util to get user site.
* Use async subprocess for one of the functions to get user site.
* Add function to package util to check if virtual environment is
  active.
* Add and update tests.

andrey-git commented:
@MartinHjelmare Martin, any updates regarding this PR?

MartinHjelmare commented:
No, not yet. I might have some time next week. Anyone is welcome to make a stab at solving the remaining things if they want.

MartinHjelmare committed:
Update version for last removal of deps dir

MartinHjelmare commented:
Ok, I've pushed the changes according to my comment above. I'm happy to take more feedback. I didn't manage to make an async pytest fixture, so the corresponding test, `test_async_get_user_site` is not so DRY.

balloob commented on homeassistant/util/package.py:
```diff
@@ -59,12 +65,43 @@ def check_package_exists(package: str, lib_dir: str) -> bool:
+    return args, env
+
+
+def get_user_site(deps_dir: str) -> str:
```
There is no reason to add sync versions if we are only going to use the async versions in our code.

MartinHjelmare replied:
It's used in `bootstrap.py` in the sync version of `mount_local_lib_path`  which in turn is used by the `scripts` package. Let me know if I should change that.

balloob commented on homeassistant/bootstrap.py:
```diff
@@ -183,7 +185,8 @@ def async_from_config_file(config_path: str,
     config_dir = os.path.abspath(os.path.dirname(config_path))
     hass.config.config_dir = config_dir
-    yield from hass.async_add_job(mount_local_lib_path, config_dir)
+    yield from hass.async_add_job(
```
You can yield from it directly.

balloob commented on homeassistant/util/package.py:
```diff
@@ -27,17 +28,22 @@ def install_package(package: str, upgrade: bool=True,
-
-        process = Popen(args, stdin=PIPE, stdout=PIPE, stderr=PIPE)
+        if target:
+            # This only works if not running in venv
```
Should we do `assert not is_virtual_env()` here?

MartinHjelmare replied:
Sounds good. Just let it blow up?

balloob commented on homeassistant/scripts/__init__.py:
```diff
@@ -40,7 +40,11 @@ def run(args: List) -> int:
     for req in getattr(script, 'REQUIREMENTS', []):
-        if not install_package(req, target=deps_dir):
+        if is_virtual_env():
+            returncode = install_package(req)
```
These should use our constraints too. (I know it's not your code…)

balloob reviewed (changes requested):
Nice! Had a few comments.

MartinHjelmare committed:
Address comments

MartinHjelmare committed:
Rewrite package util tests with pytest

* Rewrite all existing unittest class based tests for package util as
  test functions, and capitalize pytest fixtures.
* Add test for installing with target inside venv.

balloob reviewed (approved):
This is ok to merge. Since it's a big change and we're close to 0.49, I suggest we keep it for 0.50 and so won't tag it for 0.49 milestone.

MartinHjelmare commented:
Thanks, yes good to keep it for 0.50.

andrey-git commented:
Dependencies no longer install for me: 
```
2017-07-15 10:37:26 INFO (Thread-4) [homeassistant.util.package] Attempting install of paho-mqtt==1.3.0
2017-07-15 10:37:34 ERROR (Thread-4) [homeassistant.util.package] Unable to install package paho-mqtt==1.3.0: Will not install to the user site because it will lack sys.path precedence to paho-mqtt in /home/sysadmin/hass/lib/python3.5/site-packages
```

MartinHjelmare commented:
@andrey-git is `/home/sysadmin/hass` a virtual environment?

I'm thinking it's this background:
https://github.com/pypa/pip/issues/573

If so, it's bad that my check for a venv doesn't figure that out, but fixing that should fix the problem. Right now that is done like this:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/util/package.py#L73-L75

andrey-git commented:
Yes, `/home/sysadmin/hass` is venv

MartinHjelmare commented:
Latest pip is checking for a venv like this, which has another case than my function. It must be that case that returns:
```py
def running_under_virtualenv():
    """
    Return True if we're running inside a virtualenv, False otherwise.
    """
    if hasattr(sys, 'real_prefix'):
        return True
    elif sys.prefix != getattr(sys, "base_prefix", sys.prefix):
        return True

    return False
```
https://github.com/pypa/pip/blob/master/pip/locations.py#L39-L49
